policybot/pkg/testflakes/testflakes.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * Test Flakes read all rows from TestResults table in Spanner gh database.
+ * Convert them back to TestResults struct in types and bind tests based on PR
+ * number. If a test both pass and fail for the same PR, it is very likely to be flaky.
+ */
+
+package testflakes
+
+import (
+	"context"
+	"strings"
+
+	"cloud.google.com/go/spanner"
+	"google.golang.org/api/iterator"
+
+	store "istio.io/bots/policybot/pkg/storage"
+	"istio.io/pkg/log"
+)
+
+/*
+ * FlakyResult struct include the test name, whether or not the test seems to be flaky
+ * and the most recent pass and fail instance for the test.
+ */
+type FlakyResult struct {
+	TestName   string
+	PrNum      int64
+	IsFlaky    bool
+	LastPass   string
+	passResult *store.TestResult
+	LastFail   string
+	failResult *store.TestResult
+}
+
+type FlakeTest struct {
+	ctx    context.Context
+	client *spanner.Client
+	table  string
+}
+
+var scope = log.RegisterScope("TestFlaky", "Check if tests are flaky", 0)
+
+func NewFlakeTest(ctx context.Context, project string, instance string, database string, table string) (*FlakeTest, error) {
+	client, err := spanner.NewClient(ctx, "projects/"+project+"/instances/"+instance+"/databases/"+database)
+
+	if err != nil {
+		scope.Errorf(err.Error())
+		return nil, err
+	}
+
+	f := &FlakeTest{
+		ctx:    ctx,
+		client: client,
+		table:  table,
+	}
+	return f, nil
+}
+
+/*
+ * Real all rows from table in Spanner and store the results into a slice of TestResult objects.
```

Comment:
That's not a sustainable approach in the long run. Can we mark rows in the input table as "done" so that we can query for only the "not done" rows?